crates/ruff_server/src/session/workspace/ruff_settings.rs

```diff
@@ -0,0 +1,108 @@
+use ruff_linter::display_settings;
+use ruff_workspace::{
+    pyproject::settings_toml,
+    resolver::{ConfigurationTransformer, Relativity},
+};
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use walkdir::{DirEntry, WalkDir};
+
+#[derive(Default)]
+pub(crate) struct RuffSettings {
+    // settings to pass into the ruff linter
+    pub(crate) linter: ruff_linter::settings::LinterSettings,
+    // settings to pass into the ruff formatter
+    pub(crate) formatter: ruff_workspace::FormatterSettings,
+}
+
+pub(super) struct RuffSettingsIndex {
+    index: BTreeMap<PathBuf, Arc<RuffSettings>>,
+    fallback: Arc<RuffSettings>,
+}
+
+impl std::fmt::Display for RuffSettings {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        display_settings! {
+            formatter = f,
+            fields = [
+                self.linter,
+                self.formatter
+            ]
+        }
+        Ok(())
+    }
+}
+
+impl RuffSettingsIndex {
+    pub(super) fn new(path: &Path) -> Self {
+        let mut index = Self {
+            index: BTreeMap::default(),
+            fallback: Arc::default(),
+        };
+
+        index.reload(path);
+
+        index
+    }
+
+    pub(super) fn reload(&mut self, root: &Path) {
+        self.clear();
+
+        for directory in WalkDir::new(root)
+            .into_iter()
+            .filter_map(Result::ok)
+            .filter(|entry| entry.file_type().is_dir())
+            .map(DirEntry::into_path)
+        {
+            if let Some(pyproject) = settings_toml(&directory).ok().flatten() {
+                let Ok(settings) = ruff_workspace::resolver::resolve_root_settings(
+                    &pyproject,
+                    Relativity::Parent,
+                    &LSPConfigTransformer,
+                ) else {
+                    continue;
+                };
+                self.index.insert(
+                    directory,
+                    Arc::new(RuffSettings {
+                        linter: settings.linter,
+                        formatter: settings.formatter,
+                    }),
+                );
+            }
+        }
+    }
+
+    pub(super) fn get(&self, document_path: &Path) -> &Arc<RuffSettings> {
```

Comment:
Do we still need the `Arc` if we're returning a reference to `Settings`?

Reply:
We do, since callers of this method need to create a new `Arc` from its return value. Maybe I should go back to return an owned value to make this more clear.